File: course_101/exercises/easy_8/palindromic_substrings.rb
# Write a method that returns a list of all substrings of a string that are palindromic.
# That is, each substring must consist of the same sequence of characters forwards as it does backwards.
# The return value should be arranged in the same sequence as the substrings appear in the string.
# Duplicate palindromes should be included multiple times.

# You may (and should) use the substrings method you wrote in the previous exercise.

# For the purposes of this exercise, you should consider all characters and pay attention to case;
# that is, "AbcbA" is a palindrome, but neither "Abcba" nor "Abc-bA" are. In addition, assume that single characters are not palindromes.

def substrings_at_start(string)
  sub_strings = []
  1.upto(string.size) do |count|
    sub_strings << string[0, count]
  end

  sub_strings
end

def substrings(string)
  all_sub_strings = []
  0.upto(string.size - 1) do |index|
    all_sub_strings << substrings_at_start(string[index, (string.size - index)])
  end
  all_sub_strings.flatten
end

def palindrome?(string)
  string == string.reverse && string.length > 1
end

def palindromes(string)
  all_substrings = substrings(string)
  palindromes = all_substrings.select { |sub| palindrome?(sub)}
  palindromes
end

puts palindromes('abcd') == []
puts palindromes('madam') == ['madam', 'ada']
puts palindromes('hello-madam-did-madam-goodbye') == [
  'll', '-madam-', '-madam-did-madam-', 'madam', 'madam-did-madam', 'ada',
  'adam-did-mada', 'dam-did-mad', 'am-did-ma', 'm-did-m', '-did-', 'did',
  '-madam-', 'madam', 'ada', 'oo'
]
puts palindromes('knitting cassettes') == [
  'nittin', 'itti', 'tt', 'ss', 'settes', 'ette', 'tt'
]

File: course_js210_computational_thinking_and_javascript_programming/exercises/js210_small_problems/list_processing/palindromic_substrings.js
// Write a function that returns a list of all substrings of a string that are palindromic. That is, each substring must consist of the same sequence of characters forwards as backwards. The substrings in the returned list should be sorted by their order of appearance in the input string. Duplicate substrings should be included multiple times.

// You may (and should) use the substrings function you wrote in the previous exercise.

// For the purpose of this exercise, you should consider all characters and pay attention to case; that is, 'AbcbA' is a palindrome, but 'Abcba' and 'Abc-bA' are not. In addition, assume that single characters are not palindromes.

console.log(palindromes('abcd'));       // []
console.log(palindromes('madam'));      // [ "madam", "ada" ]

console.log(palindromes('hello-madam-did-madam-goodbye'));
// returns
// [ "ll", "-madam-", "-madam-did-madam-", "madam", "madam-did-madam", "ada",
//   "adam-did-mada", "dam-did-mad", "am-did-ma", "m-did-m", "-did-", "did",
//   "-madam-", "madam", "ada", "oo" ]

console.log(palindromes('knitting cassettes'));
// returns
// [ "nittin", "itti", "tt", "ss", "settes", "ette", "tt" ]


function palindromes(string) {
  return substrings(string).filter((word) => isPalindrome(word));
}

function isPalindrome(string) {
  if (string.length <= 1) return false;
  return string.split('').reverse().join('') === string;
}

function substrings(string) {
  let substrings = [];
  let substring;
  let i;
  
  for (i = 0; i < string.length; i += 1) {
    substring = substringsAtStart(string.slice(i))
    substrings = substrings.concat(substring);
  }

  return substrings;
}

function substringsAtStart(string) {
  let characters = string.split('');
  return characters.map(function (_character, index, array) {
    return array.slice(0, index + 1).join('');
  });
}
